src/Compiler/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -142,95 +235,98 @@ public CodeWriter Write(string value, int startIndex, int count)
             throw new ArgumentOutOfRangeException(nameof(startIndex));
         }
 
-        return WriteCore(value, startIndex, count);
+        return WriteCore(value.AsMemory(startIndex, count));
+    }
+
+    [InterpolatedStringHandler]
+    internal readonly ref struct CodeWriterInterpolatedStringHandler
+    {
+        private readonly CodeWriter _writer;
+
+        public CodeWriterInterpolatedStringHandler(int literalLength, int formattedCount, CodeWriter writer)
+        {
+            _writer = writer;
+        }
+
+        public void AppendLiteral(string value)
+            => _writer.Write(value);
+
+        public void AppendFormatted<T>(T value)
+        {
+            if (value is null)
+            {
+                return;
+            }
+
+            _writer.Write(value.ToString());
+        }
+    }
+
+    internal CodeWriter Write([InterpolatedStringHandlerArgument("")] ref CodeWriterInterpolatedStringHandler handler)
+    {
+        return this;
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private CodeWriter WriteCore(string value, int startIndex, int count)
+    private unsafe CodeWriter WriteCore(ReadOnlyMemory<char> chars, bool allowIndent = true)
     {
-        if (count == 0)
+        if (chars.IsEmpty)
         {
             return this;
         }
 
-        Indent(CurrentIndent);
+        if (allowIndent)
+        {
+            Indent(_indentSize);
+        }
 
-        _builder.Append(value, startIndex, count);
+        var lastChar = _lastChar;
 
-        _absoluteIndex += count;
+        AddTextChunk(chars);
 
-        // The data string might contain a partial newline where the previously
-        // written string has part of the newline.
-        var i = startIndex;
-        int? trailingPartStart = null;
+        var span = chars.Span;
 
```

Comment:
If we're designing the API, I agree. However, I'm not sure that adding a method that will currently be unused is right for this PR. Doing a FAR on `CodeWriter.Write(string)` shows that there are no callsites that pass the result of string.Format(...)`, so it really would be uncalled at the moment. Plus, I'd prefer not adding a `params` API to this type to avoid implicit allocation.